### PR TITLE
Avoid OOM in doctest CI

### PR DESCRIPTION
**Pull request by ydshieh**
# What does this PR do?

This is done by inject `gc.collect()` to the source code during pytest/doctest collecting the test to run.

### CODE REVIEW

**Comment by ydshieh** [comment]
The job runs 2x slower ...

File: src/transformers/testing_utils.py [hunk 1]
```diff
@@ -1870,9 +1870,18 @@ def preprocess_string(string, skip_cuda_tests):
         ):
             is_cuda_found = True
             break
+
     modified_string = ""
     if not is_cuda_found:
         modified_string = "".join(codeblocks)
+
+        if ">>>" in modified_string:
```
Comment (ydshieh) [inline src/transformers/testing_utils.py:1878]: only add `>>> import gc; gc.collect() # doctest: +IGNORE_RESULT` when the docstring containing `>>>`. Otherwsie too many doctest test cases and running time explodes.

File: src/transformers/testing_utils.py [hunk 2]
```diff
@@ -1870,9 +1870,18 @@ def preprocess_string(string, skip_cuda_tests):
...
+            lines = modified_string.split("\n")
+            indent = len(lines[-1]) - len(lines[-1].lstrip())
```
Comment (ydshieh) [inline src/transformers/testing_utils.py:1880]: just to add the correct indent.

**Comment by ydshieh** [comment]
Running the whole doctest: No more OOM. Only one test failure (lucky!)

Tag @sgugger so we can have doctest run until @Rocketknight1 have a complete solution in #24146.

**Review by sgugger — APPROVED** [review]
Thanks!

**Pull request merged**